kubernetes/README.md

```diff
@@ -0,0 +1,207 @@
+# torchelastic on Kubernetes
+This directory contains kubernetes resources that help users run torchelastic
+jobs on Kubernetes. We use [Amazon EKS](https://aws.amazon.com/eks/) as an example here, the steps would also work for all other cloud providers (You need to find alternative storage options in your environment)
+
+## Prerequisites
+
+1. Setup AWS credentials on your machine.
+2. [eksctl](https://eksctl.io/) to manage Kubernetes cluster.
+2. Network file storage which supports ReadWriteMany mode. [EFS](https://aws.amazon.com/efs/), [FSx for Lustrue](https://aws.amazon.com/fsx/lustre/) are good examples.
+3. [S3](https://aws.amazon.com/s3/) bucket to host training scripts.
+
+
+## Quickstart
+
+This guide shows how to get a torchelastic job running on EKS.
+
+### Create Kubernetes cluster
+
+We highly recommend to use [eksctl](https://eksctl.io/) to create Amazon EKS cluster. This process will take 10~15 minutes.
+
+``` bash
+$ eksctl create cluster \
+    --name=torchelastic \
+    --node-type=p2.xlarge \
+    --region=us-west-2 \
+    --version=1.14 \
+    --ssh-access \
+    --ssh-public-key=~/.ssh/id_rsa.pub \
+    --nodes=2
+```
+
+### Install Nvidia Device Plugin
+
+In order to enable GPU support in your EKS cluster, deploy the following Daemonset:
+
+``` bash
+$ kubectl create -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/1.0.0-beta4/nvidia-device-plugin.yml
+```
+
+### Prepare storage
+
+#### Create EFS and mount targets
+Create an EFS volume in AWS console. `eksctl` will create a new VPC for the cluster, please select the VPC to make sure EKS cluster nodes can access EFS. Select `Public` subnets when you create mount targets. Please use security group `*ClusterSharedNodeSecurityGroup*` created by `eksctl` to associate with the mount targets.
+
+![efs-setup-diagram](../docs/_static/img/efs-setup.jpg)
+
+#### Install EFS CSI Driver
+
+EFS CSI Driver manages the lifecycle of Amazon EFS filesystems in Kubernetes.
+
+``` bash
+kubectl apply -k "github.com/kubernetes-sigs/aws-efs-csi-driver/deploy/kubernetes/overlays/stable/?ref=master"
+```
+
+> Note: If you use kubernetes cluster < 1.14, please follow [instructions](https://github.com/kubernetes-sigs/aws-efs-csi-driver) to install older CSI versions.
+
+#### Create PersistentVolume and PersistentVolumeClaim
+
+Once the file system is statically provisioned and CSI driver is ready. EFS can be mounted inside a container as a volume using the driver.
+
+``` bash
+kubectl apply -f storage.yaml
+```
+
+#### Download imagenet dataset
+
+In this example, we use [tiny imagenet dataset](https://tiny-imagenet.herokuapp.com/) instead as it's a smaller dataset and easy to download.
+
+> Note: You can download full imagenet dataset [here](http://www.image-net.org/).
+
+We can create and login this pod and download dataset on the NFS.
+
+``` bash
+cat <<EOF | kubectl apply -f -
```

Comment:
[Suggestion] You can perhaps use `kubectl run` with an image like busybox so you don't need the `while` loop below.

Reply:
Got you. For simple image I agree. This one needs persistVolume support  that one line command can not handle. 
This is a copy past script and there's no `while` loop?

Reply:
Just to clarify, I was referring to line 84 where there is a `while true; do ...` loop. If there is no obvious way to eliminate it, I guess we can keep that. Agreed with your point about the PVC - looks like you do need a YAML file for that along with `kubectl apply`.

Reply:
Yeah, I got you. I remove while loop and make this as a one time task `download-dataset-task`. Pod will be complete when dataset is ready.